### ISSUE TITLE
Inline modules

### ISSUE THREAD

**Issue by iloveitaly** [casey/just#2442]
From [this discord conversation.](https://discord.com/channels/695580069837406228/695580069837406231/1300919273270673518)

I have a bunch of related tasks:

```
    db-generate-migration
    db-migrate
    db-nuke
    db-reset
    db-seed
    db-up
    db-destroy
```

What I'd like to do is have a more clear separator (like rake):

```
    db:generate-migration
    db:migrate
```

This is possible with modules, but I cannot create multiple modules in a single Justfile (creating separate Justfiles would be messy in my case).

Two ideas I can think of to solve this problem:

1. Would it be possible to support `:`, `.` in the name grammar?
2. Could we define multiple modules within a single Justfile?

**Comment by casey** [comment]
Thanks for opening this issue!

I would like to eventually support inline modules, which would allow submodules to be defined in a single justfile.

We could allow `::` in recipe names, but modules do more than just provide namespaces for recipes, since they can have their own settings, variables, and submodules, so I'm more keen to just do full blown inline modules, which would support all of that.

The syntax I've had in mind is:

```
bar::
  foo:
    echo 'recipe foo inside module bar'
```

But it could also be:

```
mod bar::
  foo:
    echo 'recipe foo inside module bar'
```

`just` uses indentation to delimit recipe bodies, since recipe bodies are arbitrary text, and `just` is descended from `make`, which does the same. So it would be natural to also use indentation to delimit modules bodies, as in the examples above.

However, I do personally find indentation to be annoying and fiddly, so we could consider using `{…}` for modules, which would allow for indenting the contents:

```
bar {
  foo:
    echo 'recipe foo inside module bar'
}
```

Or not, depending on preference:

```
bar {
foo:
  echo 'recipe foo inside module bar'
}
```

**Comment by iloveitaly** [comment]
I _really_ like the `::` indentation structure:

```
bar::
  foo:
    echo 'recipe foo inside module bar'
```

Feels really natural and easy to read.

Nested indentation isn't awesome, but `{` syntax wouldn't seem to align with the rest of the Justfile grammar IMHO. However, this is a departure from Makefile syntax in a bigger way so maybe it makes sense to do something very different?

Either way sounds great to me! I love this idea.

**Comment by casey** [comment]
Yah, honestly I like it too, especially in terms of being easy to read. My preference for avoiding indentation-based syntax is mostly when it comes to writing and editing. For example, copy-pasting a recipe from an inline module, where it's indented, to a dedicated module file, where it won't be indented, would require fixing the indentation manually.

But it's probably important that the syntax all feels consistent, and not randomly have `{…}` delimited constructs when everything else is based on indentation.

**Comment by iloveitaly** [comment]
> But it's probably important that the syntax all feels consistent

I'd agree here!

Looking forward to kicking the tires on this whenever it's implemented.

Thanks for the excellent tool.

**Comment by gechelberger** [comment]
Seconding. I was going to file this same issue.

Is the intent for these to be just logical groupings or more like clap sub commands?

I would lean more towards the subcmd option, though that is certainly more complicated:

* Can modules be nested recursively? If you have the machinery to do it at a single level, I would expect that to be most of the complexity.
* Can each module define it's own "default:" like a top level justfile
* Can a parent module have variables that are available in a child scope

```
msrv := "1.65.0"

ci opts=""::
  default: core
  core:
    cargo test {{opts}}
  msrv:
    cargo +{{msrv}} test {{opts}}
  features:
    cargo test --all-features {{opts}}
  all: features msrv
```

```
> just ci
> just ci --verbose
> just ci::features --verbose
> just ci::all --no-run
```

**Comment by casey** [comment]
They would be like subcommands, since an inline module is the same as a normal module, and normal modules create subcommands.

It should be possible to nest modules recursively, although the parsing code will likely be horrible, so it's okay if only one level of nesting is supported at first.

Every module can have its own default, like existing modules.

Modules can't share anything yet, and are a little gimped at the moment 😅, see #2252.

**Comment by iloveitaly** [comment]
After tinkering around [with an ever-growing Justfile](https://github.com/iloveitaly/python-starter-template/blob/master/Justfile), I'm going to put in my vote for:

* Allowing a single `:` in the recipe name grammar. This would still be helpful/cleaner when not using modules.
* Keeping `::` for modules

**Comment by casey** [comment]
Allowing `:` in the recipe name grammar would lead to some issues.

For example, should this:

```
foo:bar:baz
```

Be interpreted as recipe `foo` with dependency `bar:baz`, or recipe `foo:bar` with dependency `baz`. You could work around this by whitespace to disambiguate it, but that feels kinda bad.

**Comment by iloveitaly** [comment]
Ah I see, that makes sense. Any workarounds would make the syntax ugly. 

Another idea here is allowing dot-syntax. Mix does this and it works pretty well. 

`command.with.dots: other.command`

**Comment by casey** [comment]
Yah, similar issues with `.` syntax. `.` is used so common for property access, and we might want that later, i.e. `object.method()`.

**Comment by iloveitaly** [comment]
Yup yup makes sense especially if you aim to expand the language and syntax over time. Love how thoughtful you've been with the tool!

**Comment by Haramaty** [comment]
This issue has been silent for a while but if I can throw my two cents in, one of the strength of just is its readability by people who don't know the syntax. Declaring an in-line module with just `::` seems too similar to rule definition, imo. I think adding the `mod` keyword (either with `::` or `{}`) really helps with that.

**Comment by drewgingerich** [comment]
I think an upside of using `::` to define an inline module is that it's a natural connection to how a user already runs modules (with the path syntax, at least). I actually blindly tried defining an inline module using this syntax before looking around and finding this thread.

**Comment by drewgingerich** [comment]
Additional indentation and brackets could both be avoided by using some sort of ending keyword. For example `::end` or `::end <mod_name>`:

```
ci::

lint:
  yarn run lint --check
format:
  yarn run format --check
test:
  yarn run test --ci

::end ci
```

Not attached to this particular syntax, just wanted to demo the concept.


Edit: after writing the above, I thought of a different potential syntax for avoiding extra indentation. Feels like `#region` and `#endregion` from C#.

```
::start mod ci

lint:
  yarn run lint --check
format:
  yarn run format --check
test:
  yarn run test --ci

::end mod ci
```

**Comment by ozydingo** [comment]
Big plus to this whole conversation. In my case, I have a half dozen justfiles, a few for difference services in our architecture (`just app`, `just r2`, `just boba`) and some dev / setup modules `just db`, `just agents`, etc). The modules are _really nice_ (thank you!) for:

- Documentation grouping: `just --list app`
- AI agent permissions: `Allow: Bash(just app *); Deny: Bash(just agents *)`
 
The latter could be accomplished with prefixes (`app-*`), but it's awkward and potentially ambiguous while a space is an ideal delimiter here.

Inline submodules would be **great** to avoid scattering `.just` files. An alternative could be to move the justfiles into a `justfiles` folder, use an explicit path for each submodule, and use `[no-cd]` on each recipes in the submodules -- this is pretty clunky though!

```
mod foo 'justfiles/foo.just'
```

```
# justfiles/foo.just

[no-cd]
bar:
  echo bar

[no-cd]
baz:
  ech baz
```

A `no-cd` on the entire module would be slightly better

Agree with previous comments that indentation is both (a) awkward and (b) probably the best choice for consistency! I don't understand why the double `::` is necessary if we use `mod`, like

```
mod bar:
  foo:
    echo 'recipe foo inside module bar'
```

but maybe I'm missing something, and also it's fine.
